File: tests/fixtures/selenium-chaos-suite-csharp/src/Pages/CheckoutPage.cs
using OpenQA.Selenium;

namespace SeleniumChaosSuite.Pages;

public class CheckoutPage : BasePage
{
    public CheckoutPage(IWebDriver driver) : base(driver) { }

    public Task FillFirstName(string firstName) => TypeByCss("#first-name", firstName);

    public Task FillLastName(string lastName) => TypeByCss("#last-name", lastName);

    public Task FillPostalCode(string postalCode) => TypeByCss("#postal-code", postalCode);

    public async Task SubmitStepOne()
    {
        await ClickByCss("#continue");
        Wait.Until(d => d.FindElement(By.CssSelector("body")).Displayed);
    }

    public async Task FillCheckoutInfo(string firstName, string lastName, string postalCode)
    {
        await FillFirstName(firstName);
        await FillLastName(lastName);
        await FillPostalCode(postalCode);
        await SubmitStepOne();
        await AssertUrlContains("checkout-step-two");
    }

    public Task<string> GetSummaryInfo()
    {
        return GetTextByCss(".summary_info");
    }

    public async Task<string> FinishCheckout()
    {
        await ClickByCss("#finish");
        await AssertUrlContains("checkout-complete");
        return await GetTextByCss(".complete-header");
    }

    public Task<string> GetTotalLabel()
    {
        return GetTextByCss(".summary_total_label");
    }

    public async Task<string> DoBrokenSubmissionAndGetBody()
    {
        await ClickByCss("#cancel");
        Driver.Navigate().Back();
        Wait.Until(d => d.FindElement(By.CssSelector("body")).Displayed);
        await SubmitStepOne();
        return Driver.FindElement(By.CssSelector("body")).Text;
    }
}

File: tests/fixtures/selenium-chaos-suite/src/pages/CheckoutPage.ts
import { By } from "selenium-webdriver";
import { BasePage } from "./BasePage";
import { AssertionHelper } from "../core/AssertionHelper";
import { WaitUtils } from "../core/WaitUtils";

export class CheckoutPage extends BasePage {
  async fillFirstName(firstName: string) {
    await this.typeByCss("#first-name", firstName);
  }

  async fillLastName(lastName: string) {
    await this.typeByCss("#last-name", lastName);
  }

  async fillPostalCode(postalCode: string) {
    await this.typeByCss("#postal-code", postalCode);
  }

  async submitStepOne() {
    await this.clickByCss("#continue");
    await WaitUtils.sleep(300);
  }

  async fillCheckoutInfo(firstName: string, lastName: string, postalCode: string) {
    await this.fillFirstName(firstName);
    await this.fillLastName(lastName);
    await this.fillPostalCode(postalCode);
    await this.submitStepOne();
    await this.assertUrlContains("checkout-step-two");
  }

  async assertSummaryContainsPayment() {
    const info = await this.getTextByCss(".summary_info");
    AssertionHelper.shouldContain(info, "Payment Information", "Missing payment info");
    AssertionHelper.shouldContain(info, "Shipping Information", "Missing shipping info");
  }

  async finishCheckout() {
    await this.clickByCss("#finish");
    await this.assertUrlContains("checkout-complete");
    const complete = await this.getTextByCss(".complete-header");
    AssertionHelper.shouldContain(complete, "Thank you", "Completion header mismatch");
  }

  async assertTotalLooksValid() {
    const totalLine = await this.getTextByCss(".summary_total_label");
    AssertionHelper.shouldContain(totalLine, "Total:", "Total label not visible");
    AssertionHelper.shouldContain(totalLine, "$", "Total amount missing currency");
  }

  async doBrokenSubmissionAndAssertErrors() {
    await this.clickByCss("#cancel");
    await this.driver.navigate().back();
    await this.submitStepOne();
    const body = await this.driver.findElement(By.css("body")).getText();
    AssertionHelper.shouldContain(body.toLowerCase(), "error", "Expected error when fields are empty");
  }
}
